[#147552449] upgrade Ruby to 2.3

MothOnMars opened the pull request:
- upgrade to rails 5.1
- update gems
- add pessimistic version constraints to Gemfile

MothOnMars commented on config/airbrake.yml:
```diff
@@ -1 +1,2 @@
-api_key: 'set yours here'
+project_key: 'set yours here'
```
Changes required per https://github.com/airbrake/airbrake/blob/master/docs/Migration_guide_from_v4_to_v5.md. 
Should we put in a story to update the cookbooks?

noremmie replied:
I'll update the cookbooks to accommodate this change as part of my demo environment rebuild story.

MothOnMars commented on config/initializers/airbrake.rb:
```diff
@@ -1,5 +1,65 @@
 
-Airbrake.configure do |config|
-  config.api_key = yaml['api_key']
+# Airbrake is an online tool that provides robust exception tracking in your Rails
```
I regenerated this file from Airbrake 6 and updated the values accordingly.

MothOnMars commented on spec/requests/api/v1/collections_spec.rb:
```diff
@@ -13,7 +13,7 @@
         Elasticsearch::Persistence.client.delete_by_query index: Collection.index_name, q: '*:*'
         valid_params = { "handle" => "agency_blogs", "token" => "secret" }
-        post "/api/v1/collections", valid_params, valid_session
+        post "/api/v1/collections", params: valid_params, headers: valid_session
```
Spec changes required for Rails 5.1:

```
DEPRECATION WARNING: ActionDispatch::IntegrationTest HTTP request methods will accept only
the following keyword arguments in future Rails versions:
params, headers, env, xhr, as

Examples:

get '/profile',
  params: { id: 1 },
  headers: { 'X-Extra-Header' => '123' },
  env: { 'action_dispatch.custom' => 'custom' },
  xhr: true,
  as: :json
```

MothOnMars commented on app/models/application_record.rb:
```diff
@@ -0,0 +1,5 @@
+require 'active_record'
```
@noremmie , I saw your comment on this. You are correct that that line shouldn't be necessary. I am able to run `rails c` without issue in my env on this branch, without the `require`. LMK if you want to look at the error together.

noremmie replied:
@MothOnMars this is what happens when I remove that line in a demo environment:

```
[search@app1.demo3 current] (deploy)$ RAILS_ENV=production bundle exec rails c
/home/search/i14y/releases/20170704213205/app/models/application_record.rb:1:in `<top (required)>': uninitialized constant ActiveRecord (NameError)
        from /home/search/.bundler/i14y/ruby/2.3.0/gems/activesupport-5.1.1/lib/active_support/dependencies/interlock.rb:12:in `block in loading'
        from /home/search/.bundler/i14y/ruby/2.3.0/gems/activesupport-5.1.1/lib/active_support/concurrency/share_lock.rb:149:in `exclusive'
        from /home/search/.bundler/i14y/ruby/2.3.0/gems/activesupport-5.1.1/lib/active_support/dependencies/interlock.rb:11:in `loading'
        from /home/search/.bundler/i14y/ruby/2.3.0/gems/railties-5.1.1/lib/rails/engine.rb:476:in `block (2 levels) in eager_load!'
        from /home/search/.bundler/i14y/ruby/2.3.0/gems/railties-5.1.1/lib/rails/engine.rb:475:in `each'
        from /home/search/.bundler/i14y/ruby/2.3.0/gems/railties-5.1.1/lib/rails/engine.rb:475:in `block in eager_load!'
        from /home/search/.bundler/i14y/ruby/2.3.0/gems/railties-5.1.1/lib/rails/engine.rb:473:in `each'
        from /home/search/.bundler/i14y/ruby/2.3.0/gems/railties-5.1.1/lib/rails/engine.rb:473:in `eager_load!'
        from /home/search/.bundler/i14y/ruby/2.3.0/gems/railties-5.1.1/lib/rails/engine.rb:354:in `eager_load!'
        from /home/search/.bundler/i14y/ruby/2.3.0/gems/railties-5.1.1/lib/rails/application/finisher.rb:67:in `each'
        from /home/search/.bundler/i14y/ruby/2.3.0/gems/railties-5.1.1/lib/rails/application/finisher.rb:67:in `block in <module:Finisher>'
        from /home/search/.bundler/i14y/ruby/2.3.0/gems/railties-5.1.1/lib/rails/initializable.rb:30:in `instance_exec'
        from /home/search/.bundler/i14y/ruby/2.3.0/gems/railties-5.1.1/lib/rails/initializable.rb:30:in `run'
        from /home/search/.bundler/i14y/ruby/2.3.0/gems/railties-5.1.1/lib/rails/initializable.rb:59:in `block in run_initializers'
        from /usr/local/lib/ruby/2.3.0/tsort.rb:228:in `block in tsort_each'
        from /usr/local/lib/ruby/2.3.0/tsort.rb:350:in `block (2 levels) in each_strongly_connected_component'
        from /usr/local/lib/ruby/2.3.0/tsort.rb:431:in `each_strongly_connected_component_from'
        from /usr/local/lib/ruby/2.3.0/tsort.rb:349:in `block in each_strongly_connected_component'
        from /usr/local/lib/ruby/2.3.0/tsort.rb:347:in `each'
        from /usr/local/lib/ruby/2.3.0/tsort.rb:347:in `call'
        from /usr/local/lib/ruby/2.3.0/tsort.rb:347:in `each_strongly_connected_component'
        from /usr/local/lib/ruby/2.3.0/tsort.rb:226:in `tsort_each'
        from /usr/local/lib/ruby/2.3.0/tsort.rb:205:in `tsort_each'
        from /home/search/.bundler/i14y/ruby/2.3.0/gems/railties-5.1.1/lib/rails/initializable.rb:58:in `run_initializers'
        from /home/search/.bundler/i14y/ruby/2.3.0/gems/railties-5.1.1/lib/rails/application.rb:353:in `initialize!'
        from /home/search/i14y/releases/20170704213205/config/environment.rb:5:in `<top (required)>'
        from /home/search/.bundler/i14y/ruby/2.3.0/gems/railties-5.1.1/lib/rails/application.rb:329:in `require'
        from /home/search/.bundler/i14y/ruby/2.3.0/gems/railties-5.1.1/lib/rails/application.rb:329:in `require_environment!'
        from /home/search/.bundler/i14y/ruby/2.3.0/gems/railties-5.1.1/lib/rails/command/actions.rb:16:in `require_application_and_environment!'
        from /home/search/.bundler/i14y/ruby/2.3.0/gems/railties-5.1.1/lib/rails/commands/console/console_command.rb:84:in `perform'
        from /home/search/.bundler/i14y/ruby/2.3.0/gems/thor-0.19.4/lib/thor/command.rb:27:in `run'
        from /home/search/.bundler/i14y/ruby/2.3.0/gems/thor-0.19.4/lib/thor/invocation.rb:126:in `invoke_command'
        from /home/search/.bundler/i14y/ruby/2.3.0/gems/thor-0.19.4/lib/thor.rb:369:in `dispatch'
        from /home/search/.bundler/i14y/ruby/2.3.0/gems/railties-5.1.1/lib/rails/command/base.rb:63:in `perform'
        from /home/search/.bundler/i14y/ruby/2.3.0/gems/railties-5.1.1/lib/rails/command.rb:44:in `invoke'
        from /home/search/.bundler/i14y/ruby/2.3.0/gems/railties-5.1.1/lib/rails/commands.rb:16:in `<top (required)>'
        from bin/rails:4:in `require'
        from bin/rails:4:in `<main>'
```

Does any of that ring any bells with you?

MothOnMars replied:
It ain't ringing bells, but I'm curious why the directory path includes ruby `2.3.0`. Is the correct ruby version loaded (2.3.4)?

noremmie replied:
Yes, we're definitely using Ruby version 2.3.4.

```
[search@app1.demo3 current] (deploy)$ bundle env | head -7
Environment

    Bundler   1.12.0
    Rubygems  2.5.1
    Ruby      2.3.4p301 (2017-03-30 revision 58214) [x86_64-linux]
    Git       1.9.1
```

MothOnMars replied:
@noremmie - I will look into this.

MothOnMars commented:
https://github.com/GSA/i14y/pull/16/commits/332080f3adacb00b51aef0577feee17c29273b8e resolves the ActiveRecord failure - I remembered that we are not using it at all in I14y, so I removed the application_record.rb file entirely. I also added some tweaks to the Rails 5 config, generated by `rails app:update`.

noremmie commented:
Thanks, @MothOnMars. This deploys nicely now in demo environments 👍
